README
RusTFTP

A client and server implementation of the Trivial File Transfer Protocol,
written in Rust.


Currently supported:
- RFC 1350 (TFTP revision 2)
- RFC 2347 (Option Extension)
- RFC 2348 (Blocksize Option)
- RFC 2349 (Timeout Interval and Transfer Size Options)

Non-standard options:
- blksize2: block size as a power of 2
- utimeout: timeout in microseconds


Use cargo to build the binaries (output dir is target/release/):
 $ cargo build --release


Client:

 $ ./rtftpc --help
 RusTFTP
 
 ./rtftpc [options] <remote>[:port]
 
 Options:
     -h, --help          display usage information
     -g, --get FILE      download file from remote server
     -p, --put FILE      upload file to remote server
     -b, --blksize SIZE  negotiate a different block size (default: 1428)
     -n, --netascii      use netascii mode (instead of octet)


Server:

 $ ./rtftpd --help
 RusTFTP
 
 ./rtftpd [options] [directory]
 
 Options:
     -h, --help          display usage information
     -p, --port PORT     port to listen on (default: 69)
     -u, --uid UID       user id to run as (default: 65534)
     -g, --gid GID       group id to run as (default: 65534)
     -r, --read-only     allow only reading/downloading of files (RRQ)
     -w, --write-only    allow only writing/uploading of files (WRQ)
     -t, --threads N     number of worker threads (default: 2)


Notes:

 As the block number is two bytes long, the number of blocks is limited
 to 65535 (with the first block starting at 1).
 To support the transfer of files larger than 65535 blocks, it will wrap around
 after reaching the maximum and start at 0 again, which is not defined in the
 standard.


License:

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.


CHANGES
1.1.2 (2021-08-06)
  * Only non-functional changes
  * Fix new linter warnings
  * Update dependencies
  * Test also against curl in test suite
  * Add metadata for publishing to crates.io

1.1.1 (2020-01-19)
  * Skip canonical path check when running chrooted.
    This works around a requirement by musl to have a mounted /proc filesystem
    when calling realpath(3).

1.1.0 (2019-10-05)
  * Server:
    - chroot to destination directory if permissions are sufficient
    - drop -d parameter, but allow directory at the end of the command line
  * Support for some non-standard options:
    - blksize2
    - utimeout

1.0.0 (2019-03-12)
  * Initial release
  * RFC 1350 (TFTP revision 2)
  * RFC 2347 (Option Extension)
  * RFC 2348 (Blocksize Option)
  * RFC 2349 (Timeout Interval and Transfer Size Options)


test.sh
#!/bin/bash

set -e

PORT=12345
USER=$(whoami)
GROUP=$(groups | cut -d' ' -f 1)

SSD=/sbin/start-stop-daemon

RTFTPC=$(pwd)/target/release/rtftpc
RTFTPD=$(pwd)/target/release/rtftpd

ATFTPC=/usr/bin/atftp
ATFTPD=/usr/sbin/atftpd

TFTPC=/usr/bin/tftp

BUSYBOX=/bin/busybox

CURL=/usr/bin/curl

CLIENTDIR=$(mktemp -d)
SERVERDIR=$(mktemp -d)

cleanup() {
	atftpd_cleanup
	rtftpd_cleanup
	rm -f "$CLIENTDIR/testfile" "$SERVERDIR/testfile"
	rmdir "$CLIENTDIR" "$SERVERDIR"
}

compare_files() {
	cmp "$CLIENTDIR/testfile" "$SERVERDIR/testfile" 1>/dev/null
}

atftpd() {
	$ATFTPD --port $PORT --user "$USER" --group "$GROUP" --daemon "$SERVERDIR"
}

atftpc() {
	[ $TX -eq 1 ] && op="-p" || op="-g"
	[ -n "$NETASCII" ] && opts="--mode netascii"
	if [ -n "$BLKSIZE" ]; then
		$ATFTPC $op -l testfile -r testfile $opts --option "blksize $BLKSIZE" 127.0.0.1 $PORT 1>/dev/null 2>&1
	else
		$ATFTPC $op -l testfile -r testfile $opts 127.0.0.1 $PORT 1>/dev/null
	fi
}

tftpc() {
	[ $TX -eq 1 ] && op="put" || op="get"
	printf "connect 127.0.0.1 $PORT\\nmode binary\\n$op testfile\\n" | $TFTPC 1>/dev/null
}

rtftpd() {
	$SSD --background --exec "$RTFTPD" --start -- -p $PORT "$SERVERDIR" 1>/dev/null
}

rtftpc() {
	[ $TX -eq 1 ] && op="-p" || op="-g"
	[ -n "$BLKSIZE" ] && opts="--blksize $BLKSIZE"
	[ -n "$NETASCII" ] && opts="$opts -n"
	$RTFTPC $op testfile $opts 127.0.0.1:$PORT 1>/dev/null
}

curl() {
	[ $TX -eq 1 ] && op="-T" || op="-o"
	[ -n "$BLKSIZE" ] && opts="--tftp-blksize $BLKSIZE"
	$CURL --silent $op testfile $opts tftp://127.0.0.1:$PORT/testfile
}

busybox_tftpc() {
	[ $TX -eq 1 ] && op="-p" || op="-g"
	[ -n "$BLKSIZE" ] && opts="-b $BLKSIZE"
	$BUSYBOX tftp $op -l testfile -r testfile $opts 127.0.0.1 $PORT 1>/dev/null 2>&1
}

atftpd_cleanup() {
	killall -q -9 $ATFTPD 2>/dev/null || true
}

rtftpd_cleanup() {
	killall -q -9 "$RTFTPD" 2>/dev/null || true
}


test_transfer() {
	client=$1
	server=$2

	$server

	dd if=/dev/urandom of="$CLIENTDIR/testfile" bs=1M count=100 2>/dev/null

	time (
		printf "$client TX (to $server): "
		TX=1
		${client}
		compare_files
		printf "ok"
	) 2>&1
	rm -f "$CLIENTDIR/testfile"

	time (
		printf "$client RX (from $server): "
		TX=0
		${client}
		compare_files
		printf "ok"
	) 2>&1
	rm -f "$SERVERDIR/testfile"

	${server}_cleanup
}

trap cleanup 0 1 2

# make sure binaries are up-to-date
cargo build --release

cd "$CLIENTDIR"


# Defaults
printf "Testing with default configuration\\n"
test_transfer rtftpc rtftpd
[ -x $ATFTPD ]  && test_transfer rtftpc atftpd
[ -x $ATFTPC ]  && test_transfer atftpc rtftpd
[ -x $TFTPC ]   && test_transfer tftpc rtftpd
[ -x $BUSYBOX ] && test_transfer busybox_tftpc rtftpd
[ -x $CURL ]    && test_transfer curl rtftpd

# with netascii mode
printf "\\n\\nTesting netascii transfers\\n"
NETASCII=1
test_transfer rtftpc rtftpd
[ -x $ATFTPD ] && test_transfer rtftpc atftpd
[ -x $ATFTPC ] && test_transfer atftpc rtftpd
unset NETASCII

# different block size
printf "\\n\\nTesting larger block sizes\\n"
BLKSIZE=1500
test_transfer rtftpc rtftpd
[ -x $ATFTPD ]  && test_transfer rtftpc atftpd
[ -x $ATFTPC ]  && test_transfer atftpc rtftpd
[ -x $BUSYBOX ] && test_transfer busybox_tftpc rtftpd
[ -x $CURL ]    && test_transfer curl rtftpd
unset BLKSIZE

# blocksize and netascii
printf "\\n\\nTesting larger block sizes and netascii\\n"
BLKSIZE=1500
NETASCII=1
test_transfer rtftpc rtftpd
[ -x $ATFTPD ] && test_transfer rtftpc atftpd
[ -x $ATFTPC ] && test_transfer atftpc rtftpd
unset NETASCII


src/tftp.rs
/*
 * Copyright 2019 Reiner Herrmann <reiner@reiner-h.de>
 * License: GPL-3+
 */

use std::net::UdpSocket;
use std::fs::File;
use std::collections::HashMap;
use std::path::{Path,PathBuf};
use std::io;
use std::io::prelude::*;
use std::time::Duration;

pub struct TftpOptions {
    blksize: usize,
    timeout: u8,
}

pub struct Tftp {
    options: TftpOptions,
}

fn default_options() -> TftpOptions {
    TftpOptions {
        blksize: 512,
        timeout: 3,
    }
}

impl Tftp {

    pub fn new() -> Tftp {
        Tftp{
            options: default_options(),
        }
    }

    fn get_tftp_str(&self, buf: &[u8]) -> Option<(String, usize)> {
        let mut iter = buf.iter();

        let len = match iter.position(|&x| x == 0) {
            Some(l) => l,
            None => return None,
        };
        let val = match String::from_utf8(buf[0 .. len].to_vec()) {
            Ok(v) => v,
            Err(_) => return None,
        };

        return Some((val, len));
    }

    fn wait_for_ack(&self, sock: &UdpSocket, expected_block: u16) -> Result<bool, io::Error> {
        let mut buf = [0; 4];
        match sock.recv(&mut buf) {
            Ok(_) => (),
            Err(ref error) if [io::ErrorKind::WouldBlock, io::ErrorKind::TimedOut].contains(&error.kind()) => {
                return Ok(false);
            }
            Err(err) => return Err(err),
        };

        let opcode = u16::from_be_bytes([buf[0], buf[1]]);
        let block_nr = u16::from_be_bytes([buf[2], buf[3]]);

        if opcode == 4 && block_nr == expected_block {
            return Ok(true)
        }

        Ok(false)
    }

    fn ack_options(&self, sock: &UdpSocket, options: &HashMap<String, String>, ackwait: bool) -> Result<(), io::Error> {
        if options.is_empty() {
            if !ackwait {
                /* it's a WRQ, send normal ack to start transfer */
                self.send_ack(&sock, 0)?;
            }
            return Ok(())
        }

        let mut buf = Vec::with_capacity(512);
        buf.extend([0x00, 0x06].iter());  // opcode

        for (key, val) in options {
            buf.extend(key.bytes());
            buf.push(0x00);
            buf.extend(val.bytes());
            buf.push(0x00);
        }

        for _ in 1..5 {
            sock.send(&buf)?;
            if !ackwait {
                return Ok(());
            }
            match self.wait_for_ack(&sock, 0) {
                Ok(true) => return Ok(()),
                Ok(false) => continue,
                Err(e) => return Err(e),
            };
        }

        Err(io::Error::new(io::ErrorKind::TimedOut, "ack timeout"))
    }

    pub fn init_tftp_options(&mut self, sock: &UdpSocket, options: &mut HashMap<String, String>, ackwait: bool) -> Result<(), io::Error> {
        self.options = default_options();

        options.retain(|key, val| {
            let val = val.to_lowercase();
            match key.to_lowercase().as_str() {
                "blksize" => {
                    match val.parse() {
                        Ok(b) if b >= 8 && b <= 65464 => {
                            self.options.blksize = b;
                            true
                        }
                        _ => false
                    }
                }
                "timeout" => {
                    match val.parse() {
                        Ok(t) if t >= 1 => {
                            self.options.timeout = t;
                            true
                        }
                        _ => false
                    }
                }
                _ => false
            }
        });

        sock.set_read_timeout(Some(Duration::from_secs(self.options.timeout as u64)))?;

        self.ack_options(&sock, &options, ackwait)?;

        return Ok(());
    }

    fn parse_options(&self, buf: &[u8]) -> HashMap<String, String> {
        let mut options = HashMap::new();

        let mut pos = 0;
        loop {
            let (key, len) = match self.get_tftp_str(&buf[pos ..]) {
                Some(args) => args,
                None => break,
            };
            pos += len + 1;

            let (val, len) = match self.get_tftp_str(&buf[pos ..]) {
                Some(args) => args,
                None => break,
            };
            pos += len + 1;

            options.insert(key, val);
        }

        return options;
    }

    pub fn parse_file_mode_options(&self, buf: &[u8]) -> Result<(PathBuf, String, HashMap<String, String>), io::Error> {
        let dataerr = io::Error::new(io::ErrorKind::InvalidData, "invalid data received");

        let mut pos = 0;
        let (filename, len) = match self.get_tftp_str(&buf[pos ..]) {
            Some(args) => args,
            None => return Err(dataerr),
        };
        pos += len + 1;

        let filename = Path::new(&filename);

        let (mode, len) = match self.get_tftp_str(&buf[pos ..]) {
            Some((m,l)) => (m.to_lowercase(), l),
            None => return Err(dataerr),
        };
        pos += len + 1;

        let options = self.parse_options(&buf[pos ..]);

        Ok((filename.to_path_buf(), mode, options))
    }

    pub fn send_error(&self, socket: &UdpSocket, code: u16, msg: &str) -> Result<(), io::Error> {
        let mut buf = vec![0x00, 0x05];  // opcode
        buf.extend(code.to_be_bytes().iter());
        buf.extend(msg.as_bytes());

        socket.send(&buf)?;
        Ok(())
    }

    pub fn send_ack(&self, sock: &UdpSocket, block_nr: u16) -> Result<(), io::Error> {
        let mut buf = vec![0x00, 0x04];  // opcode
        buf.extend(block_nr.to_be_bytes().iter());

        sock.send(&buf)?;

        Ok(())
    }

    pub fn send_file(&self, socket: &UdpSocket, file: &mut File) -> Result<(), io::Error> {
        let mut block_nr: u16 = 1;

        loop {
            let mut filebuf = vec![0; self.options.blksize];
            let len = match file.read(&mut filebuf) {
                Ok(n) => n,
                Err(ref error) if error.kind() == io::ErrorKind::Interrupted => continue, /* retry */
                Err(err) => {
                    self.send_error(&socket, 0, "File reading error")?;
                    return Err(err);
                }
            };

            let mut sendbuf = Vec::with_capacity(4 + len);
            sendbuf.extend([0x00, 0x03].iter());  // opcode
            sendbuf.extend(block_nr.to_be_bytes().iter());
            sendbuf.extend(filebuf[0..len].iter());

            let mut acked = false;
            for _ in 1..5 {
                /* try a couple of times to send data, in case of timeouts
                   or re-ack of previous data */
                socket.send(&sendbuf)?;
                match self.wait_for_ack(&socket, block_nr) {
                    Ok(true) => {
                        acked = true;
                        break;
                    },
                    Ok(false) => continue,
                    Err(e) => return Err(e),
                };
            }
            if !acked {
                return Err(io::Error::new(io::ErrorKind::TimedOut, "ack timeout"))
            }

            if len < self.options.blksize {
                /* this was the last block */
                break;
            }

            /* increment with rollover on overflow */
            block_nr = block_nr.wrapping_add(1);
        }
        Ok(())
    }

    pub fn recv_file(&self, sock: &UdpSocket, file: &mut File) -> Result<(), io::Error> {
        let mut block_nr: u16 = 1;

        loop {
            let mut buf = vec![0; 4 + self.options.blksize + 1]; // +1 for later size check
            let mut len = 0;

            for _ in 1..5 {
                len = match sock.recv(&mut buf) {
                    Ok(n) => n,
                    Err(ref error) if [io::ErrorKind::WouldBlock, io::ErrorKind::TimedOut].contains(&error.kind()) => {
                        /* re-ack previous and try to recv again */
                        self.send_ack(&sock, block_nr - 1)?;
                        continue;
                    }
                    Err(err) => return Err(err),
                };
                break;
            }
            if len < 4 || len > 4 + self.options.blksize {
                /* max size: 2 + 2 + blksize */
                return Err(io::Error::new(io::ErrorKind::InvalidInput, "unexpected size"));
            }

            let _opcode = match u16::from_be_bytes([buf[0], buf[1]]) {
                3 /* DATA */ => (),
                _ => return Err(io::Error::new(io::ErrorKind::Other, "unexpected opcode")),
            };
            if u16::from_be_bytes([buf[2], buf[3]]) != block_nr {
                /* already received or packets were missed, re-acknowledge */
                self.send_ack(&sock, block_nr - 1)?;
                continue;
            }

            let databuf = &buf[4..len];
            file.write_all(databuf)?;

            self.send_ack(&sock, block_nr)?;
            block_nr = block_nr.wrapping_add(1);

            if len < 4 + self.options.blksize {
                break;
            }
        }

        file.flush()?;

        Ok(())
    }

}


src/tftpd.rs
/*
 * Copyright 2019-2020 Reiner Herrmann <reiner@reiner-h.de>
 * License: GPL-3+
 */

use std::env;
use std::error::Error;
use std::fs::File;
use std::fs::OpenOptions;
use std::io;
use std::net::{SocketAddr, UdpSocket};
use std::path::{Path, PathBuf};
use std::time::Duration;

extern crate nix;
use nix::unistd::{chroot, setresgid, setresuid, Gid, Uid, ROOT};

extern crate getopts;
use getopts::Options;

extern crate threadpool;
use threadpool::ThreadPool;

extern crate rtftp;

#[derive(Clone)]
struct Configuration {
    port: u16,
    uid: u32,
    gid: u32,
    ro: bool,
    wo: bool,
    threads: usize,
    dir: PathBuf,
}

impl Default for Configuration {
    fn default() -> Configuration {
        Configuration {
            port: 69,
            uid: 65534,
            gid: 65534,
            ro: false,
            wo: false,
            threads: 2,
            dir: env::current_dir().expect("Can't get current directory"),
        }
    }
}

#[derive(Clone)]
struct Tftpd {
    tftp: rtftp::Tftp,
    conf: Configuration,
}

impl Tftpd {
    pub fn new(conf: Configuration) -> Tftpd {
        Tftpd {
            tftp: rtftp::Tftp::new(),
            conf,
        }
    }

    fn file_allowed(&self, filename: &Path) -> Option<PathBuf> {
        if self.conf.dir == PathBuf::from("/") {
            /* running either chrooted in requested directory,
               or whole root is being served */
            return Some(filename.to_path_buf());
        }

        /* get parent to check dir where file should be read/written */
        let path = self.conf.dir.join(filename)
                                .parent()?
                                .canonicalize()
                                .ok()?;

        /* check last component of given filename appended to canonicalized path */
        match path.join(filename.file_name()?).strip_prefix(&self.conf.dir) {
            Ok(p) if p != PathBuf::new() => Some(p.to_path_buf()),
            _ => None,
        }
    }

    fn handle_wrq(&mut self, socket: &UdpSocket, cl: &SocketAddr, buf: &[u8]) -> Result<String, io::Error> {
        let (filename, mode, mut options) = self.tftp.parse_file_mode_options(buf)?;
        self.tftp.init_tftp_options(socket, &mut options)?;

        match mode.as_ref() {
            "octet" => self.tftp.set_mode(rtftp::Mode::OCTET),
            "netascii" => self.tftp.set_mode(rtftp::Mode::NETASCII),
            _ => {
                self.tftp.send_error(socket, 0, "Unsupported mode")?;
                return Err(io::Error::new(io::ErrorKind::Other, "unsupported mode"));
            }
        }

        let path = match self.file_allowed(&filename) {
            Some(p) => p,
            None => {
                let err = format!("Receiving {} from {} failed (permission check failed).", filename.display(), cl);
                self.tftp.send_error(socket, 2, "Permission denied")?;
                return Err(io::Error::new(io::ErrorKind::PermissionDenied, err));
            }
        };

        let mut file = match OpenOptions::new().write(true).create_new(true).open(&path) {
            Ok(f) => f,
            Err(ref err) if err.kind() == io::ErrorKind::AlreadyExists => {
                let error = format!("Receiving {} from {} failed ({}).", path.display(), cl, err);
                self.tftp.send_error(socket, 6, "File already exists")?;
                return Err(io::Error::new(err.kind(), error));
            }
            Err(err) => {
                let error = format!("Receiving {} from {} failed ({}).", path.display(), cl, err);
                self.tftp.send_error(socket, 6, "Permission denied")?;
                return Err(io::Error::new(io::ErrorKind::PermissionDenied, error));
            }
        };

        self.tftp.ack_options(socket, &options, false)?;
        match self.tftp.recv_file(socket, &mut file) {
            Ok(_) => Ok(format!("Received {} from {}.", path.display(), cl)),
            Err(ref err) => {
                let error = format!("Receiving {} from {} failed ({}).", path.display(), cl, err);
                self.tftp.send_error(socket, 0, "Receiving error")?;
                Err(io::Error::new(err.kind(), error))
            }
        }
    }

    fn handle_rrq(&mut self, socket: &UdpSocket, cl: &SocketAddr, buf: &[u8]) -> Result<String, io::Error> {
        let (filename, mode, mut options) = self.tftp.parse_file_mode_options(buf)?;
        self.tftp.init_tftp_options(socket, &mut options)?;

        match mode.as_ref() {
            "octet" => self.tftp.set_mode(rtftp::Mode::OCTET),
            "netascii" => self.tftp.set_mode(rtftp::Mode::NETASCII),
            _ => {
                self.tftp.send_error(socket, 0, "Unsupported mode")?;
                return Err(io::Error::new(io::ErrorKind::Other, "unsupported mode"));
            }
        }

        let path = match self.file_allowed(&filename) {
            Some(p) => p,
            None => {
                let err = format!("Sending {} to {} failed (permission check failed).", filename.display(), cl);
                self.tftp.send_error(socket, 2, "Permission denied")?;
                return Err(io::Error::new(io::ErrorKind::PermissionDenied, err));
            }
        };

        let mut file = match File::open(&path) {
            Ok(f) => f,
            Err(ref error) if error.kind() == io::ErrorKind::NotFound => {
                let err = format!("Sending {} to {} failed ({}).", path.display(), cl, error.to_string());
                self.tftp.send_error(socket, 1, "File not found")?;
                return Err(io::Error::new(io::ErrorKind::NotFound, err));
            }
            Err(error) => {
                let err = format!("Sending {} to {} failed ({}).", path.display(), cl, error.to_string());
                self.tftp.send_error(socket, 2, "Permission denied")?;
                return Err(io::Error::new(io::ErrorKind::PermissionDenied, err));
            }
        };
        if !file.metadata()?.is_file() {
            self.tftp.send_error(socket, 1, "File not found")?;
            return Err(io::Error::new(io::ErrorKind::NotFound, "file not found"));
        }

        if let Some(opt) = options.get_mut("tsize") {
            *opt = self.tftp.transfersize(&mut file)?.to_string();
        }
        self.tftp.ack_options(socket, &options, true)?;
        match self.tftp.send_file(socket, &mut file) {
            Ok(_) => Ok(format!("Sent {} to {}.", path.display(), cl)),
            Err(err) => {
                let error = format!("Sending {} to {} failed ({}).", path.display(), cl, err.to_string());
                Err(std::io::Error::new(err.kind(), error))
            }
        }
    }

    pub fn handle_client(&mut self, cl: &SocketAddr, buf: &[u8]) -> Result<String, io::Error> {
        let socket = UdpSocket::bind("[::]:0")?;
        socket.set_read_timeout(Some(Duration::from_secs(5)))?;
        socket.connect(cl)?;

        if buf.len() < 2 {
            self.tftp.send_error(&socket, 0, "Invalid request length")?;
            return Err(io::Error::new(io::ErrorKind::Other, "invalid request length"));
        }

        match u16::from_be_bytes([buf[0], buf[1]]) {  // opcode
            o if o == rtftp::Opcode::RRQ as u16 => {
                if self.conf.wo {
                    self.tftp.send_error(&socket, 4, "reading not allowed")?;
                    Err(io::Error::new(io::ErrorKind::Other, "unallowed mode"))
                } else {
                    self.handle_rrq(&socket, cl, &buf[2..])
                }
            }
            o if o == rtftp::Opcode::WRQ as u16 => {
                if self.conf.ro {
                    self.tftp.send_error(&socket, 4, "writing not allowed")?;
                    Err(io::Error::new(io::ErrorKind::Other, "unallowed mode"))
                } else {
                    self.handle_wrq(&socket, cl, &buf[2..])
                }
            }
            o if o == rtftp::Opcode::ERROR as u16 => Ok(format!("Received ERROR from {}", cl)),
            _ => {
                self.tftp.send_error(&socket, 4, "Unexpected opcode")?;
                Err(io::Error::new(io::ErrorKind::Other, "unexpected opcode"))
            }
        }
    }

    fn drop_privs(&self, uid: u32, gid: u32) -> Result<(), Box<dyn Error>> {
        let root_uid = ROOT;
        let root_gid = Gid::from_raw(0);
        let unpriv_uid = Uid::from_raw(uid);
        let unpriv_gid = Gid::from_raw(gid);

        if Gid::current() != root_gid
            && Gid::effective() != root_gid
            && Uid::current() != root_uid
            && Uid::effective() != root_uid
        {
            /* already unprivileged user */
            return Ok(());
        }

        if Gid::current() == root_gid || Gid::effective() == root_gid {
            setresgid(unpriv_gid, unpriv_gid, unpriv_gid)?;
        }

        if Uid::current() == root_uid || Uid::effective() == root_uid {
            setresuid(unpriv_uid, unpriv_uid, unpriv_uid)?;
        }

        Ok(())
    }

    fn chroot_destdir(&mut self) -> Result<(), nix::Error> {
        /* chroot will only succeed if we have required permissions;
           either running as root or having CAP_SYS_CHROOT.
           propagate error only if chroot should have succeeded. */
        match chroot(&self.conf.dir) {
            Ok(_) => {
                /* configured dir is now new root directory */
                self.conf.dir = PathBuf::from("/");
                Ok(())
            },
            Err(err) if err == nix::errno::Errno::EPERM => Ok(()),
            Err(err) if Uid::effective() == ROOT => Err(err),
            Err(_) => Ok(()),
        }
    }

    pub fn start(&mut self) {
        let socket = match UdpSocket::bind(format!("[::]:{}", self.conf.port)) {
            Ok(s) => s,
            Err(err) => {
                eprintln!("Binding a socket failed: {}", err);
                return;
            }
        };
        match self.chroot_destdir() {
            Ok(_) => {},
            Err(err) => {
                eprintln!("Changing root directory failed ({}).", err);
                return;
            }
        }
        match self.drop_privs(self.conf.uid, self.conf.gid) {
            Ok(_) => (),
            Err(err) => {
                eprintln!("Dropping privileges failed: {}", err);
                return;
            }
        };

        match env::set_current_dir(&self.conf.dir) {
            Ok(_) => (),
            Err(err) => {
                eprintln!("Changing directory failed ({}).", err);
                return;
            }
        }

        let pool = ThreadPool::new(self.conf.threads);
        loop {
            let mut buf = [0; 2048];
            let (n, src) = match socket.recv_from(&mut buf) {
                Ok(args) => args,
                Err(err) => {
                    eprintln!("Receiving data from socket failed: {}", err);
                    break;
                }
            };

            let mut worker = self.clone();
            pool.execute(move || {
                match worker.handle_client(&src, &buf[0..n]) {
                    Ok(msg) => println!("{}", msg),
                    Err(err) => println!("{}", err),
                }
            });
        }
    }
}

fn usage(opts: &Options, program: &str, error: Option<String>) {
    if let Some(err) = error {
        println!("{}\n", err);
    }
    let version = rtftp::VERSION.unwrap_or("");
    println!("{}", opts.usage(format!("RusTFTP {}\n\n{} [options] [directory]", version, program).as_str()));
}

fn parse_commandline(args: &[String]) -> Option<Configuration> {
    let program = args[0].clone();
    let mut conf: Configuration = Default::default();
    let mut opts = Options::new();
    opts.optflag("h", "help", "display usage information");
    opts.optopt("p", "port", format!("port to listen on (default: {})", conf.port).as_ref(), "PORT");
    opts.optopt("u", "uid", format!("user id to run as (default: {})", conf.uid).as_ref(), "UID");
    opts.optopt("g", "gid", format!("group id to run as (default: {})", conf.gid).as_ref(), "GID");
    opts.optflag("r", "read-only", "allow only reading/downloading of files (RRQ)");
    opts.optflag("w", "write-only", "allow only writing/uploading of files (WRQ)");
    opts.optopt("t", "threads", format!("number of worker threads (default: {})", conf.threads).as_ref(), "N");

    let getopts_fail = |err: getopts::Fail| { usage(&opts, &program, Some(err.to_string())) };
    let conv_error = |err: std::num::ParseIntError| { usage(&opts, &program, Some(err.to_string())) };

    let matches = opts.parse(&args[1..]).map_err(getopts_fail).ok()?;
    if matches.opt_present("h") {
        usage(&opts, &program, None);
        return None;
    }

    conf.port = matches.opt_get_default("p", conf.port).map_err(conv_error).ok()?;
    conf.uid = matches.opt_get_default("u", conf.uid).map_err(conv_error).ok()?;
    conf.gid = matches.opt_get_default("g", conf.gid).map_err(conv_error).ok()?;
    conf.threads = matches.opt_get_default("t", conf.threads).map_err(conv_error).ok()?;
    conf.ro = matches.opt_present("r");
    conf.wo = matches.opt_present("w");
    if conf.ro && conf.wo {
        usage(&opts, &program, Some(String::from("Only one of r (read-only) and w (write-only) allowed")));
        return None;
    }
    if !matches.free.is_empty() {
        conf.dir = Path::new(&matches.free[0]).to_path_buf();
    }

    Some(conf)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let conf = match parse_commandline(&args) {
        Some(c) => c,
        None => return,
    };

    Tftpd::new(conf).start();
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_file_allowed() {
        let conf: Configuration = Default::default();
        let tftpd = Tftpd::new(conf);

        /* allowed */
        assert!(tftpd.file_allowed(Path::new("testfile")).is_some());
        assert!(tftpd.file_allowed(&tftpd.conf.dir.join(Path::new("testfile"))).is_some());

        /* forbidden */
        assert!(tftpd.file_allowed(Path::new("nonexisting_dir/testfile")).is_none());
        assert!(tftpd.file_allowed(Path::new("/nonexisting_dir/testfile")).is_none());
        assert!(tftpd.file_allowed(Path::new("../testfile")).is_none());
        assert!(tftpd.file_allowed(Path::new("testfile/../")).is_none());
        assert!(tftpd.file_allowed(Path::new("testfile/../testfile")).is_none());
        assert!(tftpd.file_allowed(Path::new("/root/testfile")).is_none());
        assert!(tftpd.file_allowed(Path::new("/testfile")).is_none());
        assert!(tftpd.file_allowed(Path::new("/dev/null")).is_none());
        assert!(tftpd.file_allowed(Path::new("../../../../../../../../../../../../../etc/motd")).is_none());
        assert!(tftpd.file_allowed(Path::new("")).is_none());
        assert!(tftpd.file_allowed(Path::new("./")).is_none());
        assert!(tftpd.file_allowed(&tftpd.conf.dir).is_none());
    }
}
